File: android/core/src/test/kotlin/com/stockbook/core/LoanTests.kt
package com.stockbook.core

import com.stockbook.core.model.Loan
import com.stockbook.core.model.Statement
import com.stockbook.core.model.StatementPeriod
import com.stockbook.core.store.DayEntryKind
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.transfer.BackupService
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.Strings
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Cash the shop lends a customer.
 *
 * Half of what is asserted here is that a loan is **not** the other things it
 * resembles — not a sale, not an expense, not a payment with a sign on it. A rule
 * of the form "these two must never meet" decays silently, which is why it is
 * written down as a test rather than left in a comment.
 */
class LoanTests {

    private val day: Instant = Instant.parse("2026-09-08T09:00:00Z")
    private val strings = Strings(AppLanguage.ENGLISH)

    private fun shop(): StockbookStore {
        val store = StockbookStore(InMemoryRepository())
        store.addCustomer("Ahmed")
        return store
    }

    // --- What it does to the balance

    /**
     * The whole feature in one assertion: lending 400 to somebody owing 800
     * leaves them owing 1,200.
     *
     * One debt, not two. Goods owed and cash lent land on the same balance,
     * because there is no second balance anywhere in this app to put it in and
     * an owner asking what Ahmed owes wants one number.
     */
    @Test
    fun `a loan adds to what the customer owes`() {
        val store = shop()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 800.0, createdAt = day)

        store.recordLoan("ahmed", 400.0, lentAt = day)

        assertEquals(1_200.0, store.customer("ahmed")?.owed)
    }

    /** Repayment is an ordinary payment. Nothing new was needed for it. */
    @Test
    fun `an ordinary payment settles a loan`() {
        val store = shop()
        store.recordLoan("ahmed", 400.0, lentAt = day)

        store.recordPayment("ahmed", 400.0, receivedAt = day.plusSeconds(86_400))

        assertEquals(0.0, store.customer("ahmed")?.owed)
    }

    /**
     * Somebody the owner has only ever lent money to still appears.
     *
     * The seeding trap: applying loans with `book[key]?.let` alone would drop a
     * customer who has no bill and no roster entry, without a sound. That is not
     * a rare case — it is the first loan to a new face, and it is exactly how the
     * credit notes were once stranded by a rename.
     */
    @Test
    fun `a customer known only by a loan is still on the books`() {
        val store = StockbookStore(InMemoryRepository())

        store.recordLoan("khalid", 250.0, lentAt = day)

        val khalid = store.customers().firstOrNull { it.key == "khalid" }
        assertNotNull(khalid, "somebody who has only ever borrowed is still a customer")
        assertEquals(250.0, khalid.owed)
    }

    // --- What it is not

    /**
     * Not a sale. Nothing left the shelf and the shop earned nothing.
     *
     * Half of `ExpenseTests` asserts absences for this reason and so does this:
     * a loan that crept into takings would flatter every month it happened in.
     */
    @Test
    fun `a loan is not a sale and earns nothing`() {
        val store = shop()
        val period = StatementPeriod.Month(day)

        store.recordLoan("ahmed", 400.0, lentAt = day)

        assertEquals(0.0, store.soldIn(period), "not takings")
        assertEquals(0.0, store.earningsIn(period).sold, "not revenue")
        assertEquals(0.0, store.spentIn(period), "and not the owner's spending either")
        assertEquals(0, store.billCountIn(period), "no bill was written")
    }

    /**
     * Not an expense. An expense is gone; this is expected back.
     *
     * They are both money out and they must not be added together anywhere — the
     * shop that treated a loan as spending would report a month's costs it never
     * had.
     */
    @Test
    fun `lending stands apart from spending`() {
        val store = shop()
        val period = StatementPeriod.Month(day)
        store.addExpense(60.0, "Petrol", day)
        store.recordLoan("ahmed", 400.0, lentAt = day)

        assertEquals(60.0, store.spentIn(period))
        assertEquals(400.0, store.lentIn(period))
    }

    /** Neither zero nor a negative figure is a loan. A negative one is a payment. */
    @Test
    fun `a loan has to be money, and has to be for somebody`() {
        val store = shop()

        assertNull(store.recordLoan("ahmed", 0.0, lentAt = day))
        assertNull(store.recordLoan("ahmed", -400.0, lentAt = day))
        assertNull(store.recordLoan("", 400.0, lentAt = day))
        assertTrue(store.loans.isEmpty())
    }

    // --- On the statement

    /**
     * A loan charges the account and settles nothing — the mirror of a payment,
     * which settles and charges nothing.
     */
    @Test
    fun `the statement shows a loan as a charge that settles nothing`() {
        val store = shop()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 800.0, createdAt = day)
        store.recordLoan("ahmed", 400.0, lentAt = day.plusSeconds(3_600))

        val statement = store.statementForCustomer("ahmed", StatementPeriod.Month(day))
        assertNotNull(statement)

        val loan = statement.entries.filterIsInstance<Statement.Entry.ForLoan>().single()
        assertEquals(400.0, loan.charge)
        assertEquals(0.0, loan.settledAtOnce)
        assertEquals(1_200.0, statement.closingBalance)
    }

    /**
     * **Billed says what was billed.** Eight hundred of goods and four hundred
     * lent is not "billed 1,200" — the customer never got an invoice for the
     * cash, and would be right to query one.
     *
     * This is why a loan is its own `Entry.Kind` rather than another TRADE entry.
     */
    @Test
    fun `a loan is kept out of what was billed`() {
        val store = shop()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 800.0, createdAt = day)
        store.recordLoan("ahmed", 400.0, lentAt = day.plusSeconds(3_600))

        val statement = store.statementForCustomer("ahmed", StatementPeriod.Month(day))!!

        assertEquals(800.0, statement.billed, "goods only")
        assertEquals(400.0, statement.lent, "and the cash on its own line")
        assertEquals(0.0, statement.received)
    }

    /** No number to print, because a loan comes out of no numbered book. */
    @Test
    fun `a loan row is named rather than numbered`() {
        val store = shop()
        store.recordLoan("ahmed", 400.0, lentAt = day)
        val statement = store.statementForCustomer("ahmed", StatementPeriod.Month(day))!!

        val row = statement.entries.single()
        assertEquals("Loan", com.stockbook.core.text.StatementDocument.reference(row, strings))
    }

    // --- On the day

    /**
     * Money out of the cash box on the day it left it, however certain the owner
     * is of getting it back. The day's net is what the box did.
     */
    @Test
    fun `a loan is money out on the day it was handed over`() {
        val store = shop()
        store.recordPayment("ahmed", 100.0, receivedAt = day)
        store.recordLoan("ahmed", 400.0, lentAt = day.plusSeconds(60))

        val book = store.dayBook(day)

        assertTrue(DayEntryKind.LOAN in book.entries.map { it.kind })
        assertEquals(100.0, book.moneyIn)
        assertEquals(400.0, book.moneyOut)
        assertEquals(-300.0, book.net)
    }

    // --- Correcting one

    @Test
    fun `a loan can be corrected and removed, and the balance follows`() {
        val store = shop()
        val loan = store.recordLoan("ahmed", 400.0, lentAt = day)!!

        store.updateLoan(loan.id, amount = 250.0, lentAt = day)
        assertEquals(250.0, store.customer("ahmed")?.owed)

        store.deleteLoan(loan.id)
        assertEquals(0.0, store.customer("ahmed")?.owed)
        assertTrue(store.loans.isEmpty())
    }

    // --- Across the wire

    /**
     * A loan survives export and import, and the file says version 5.
     *
     * The bump is the point: a reader built before loans existed would drop them
     * and show every borrower owing less than they do, so it must refuse the file
     * instead. That is the credit-note rule pointing the other way.
     */
    @Test
    fun `loans travel in the backup, which is version 5`() {
        val store = shop()
        store.saveBill(customer = "Ahmed", paid = 0.0, amount = 800.0, createdAt = day)
        store.recordLoan("ahmed", 400.0, lentAt = day, note = "for the school fees")

        val document = store.makeBackupDocument(day)
        assertEquals(5, document.version)

        val restored = StockbookStore(InMemoryRepository())
        restored.replaceEverything(BackupService.decode(BackupService.encode(document)))

        val loan: Loan = restored.loans.single()
        assertEquals(400.0, loan.amount)
        assertEquals("ahmed", loan.customerKey)
        assertEquals("for the school fees", loan.note)
        assertEquals(1_200.0, restored.customer("ahmed")?.owed, "and the balance comes back with it")
    }

    // --- Finding one

    @Test
    fun `a loan can be searched for by the borrower's name`() {
        val store = shop()
        store.recordLoan("ahmed", 400.0, lentAt = day)

        val hit = store.search("ahmed").single { it.kind == DayEntryKind.LOAN }

        assertEquals(400.0, hit.amount)
        assertEquals("Ahmed", hit.who)
        assertNull(hit.reference, "there is no number to show")
    }
}

File: ios/StockbookTests/LoanTests.swift
import Testing
import Foundation
@testable import Stockbook

/// Cash the shop lends a customer.
///
/// The twin of `LoanTests.kt`, test for test. Half of what is asserted here is
/// that a loan is **not** the other things it resembles — not a sale, not an
/// expense, not a payment with a sign on it. A rule of the form "these two must
/// never meet" decays silently, which is why it is written down as a test rather
/// than left in a comment.
@Suite("Loans")
@MainActor
struct LoanTests {

    private let english = Strings(language: .english)

    /// A fixed date rather than an offset from now, so the month boundaries
    /// these tests lean on cannot drift with the day the suite is run.
    private var day: Date { Date(timeIntervalSince1970: 1_789_203_600) }

    private func makeStore() -> StockbookStore {
        let store = StockbookStore(repository: InMemoryRepository())
        _ = store.addCustomer(name: "Ahmed")
        return store
    }

    // MARK: What it does to the balance

    /// The whole feature in one assertion: lending 400 to somebody owing 800
    /// leaves them owing 1,200.
    ///
    /// One debt, not two. Goods owed and cash lent land on the same balance,
    /// because there is no second balance anywhere in this app to put it in and
    /// an owner asking what Ahmed owes wants one number.
    @Test func aLoanAddsToWhatTheCustomerOwes() throws {
        let store = makeStore()
        _ = store.saveBill(customer: "Ahmed", paid: 0, amount: 800, createdAt: day)

        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        #expect(store.customer(key: "ahmed")?.owed == 1_200)
    }

    /// Repayment is an ordinary payment. Nothing new was needed for it.
    @Test func anOrdinaryPaymentSettlesALoan() throws {
        let store = makeStore()
        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        _ = store.recordPayment(
            customerKey: "ahmed", amount: 400, receivedAt: day.addingTimeInterval(86_400)
        )

        #expect(store.customer(key: "ahmed")?.owed == 0)
    }

    /// Somebody the owner has only ever lent money to still appears.
    ///
    /// The seeding trap: applying loans to the book without seeding it first
    /// drops a customer who has no bill and no roster entry, in silence. That is
    /// not a rare case — it is the first loan to a new face, and it is exactly
    /// how the credit notes were once stranded by a rename.
    @Test func aCustomerKnownOnlyByALoanIsStillOnTheBooks() throws {
        let store = StockbookStore(repository: InMemoryRepository())

        _ = store.recordLoan(customerKey: "khalid", amount: 250, lentAt: day)

        let khalid = try #require(store.customers().first { $0.key == "khalid" })
        #expect(khalid.owed == 250)
    }

    /// A renamed borrower keeps their debt.
    ///
    /// The failure this prevents is silent and one-directional: a loan left under
    /// the old key stops adding to what the customer owes, so the balance falls
    /// by whatever they were lent and nothing anywhere says why.
    @Test func renamingABorrowerCarriesTheirLoans() throws {
        let store = makeStore()
        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        _ = store.updateCustomer(key: "ahmed", name: "Ahmed Al Faisal", phone: nil, place: nil)

        #expect(store.customer(key: "ahmed al faisal")?.owed == 400)
    }

    // MARK: What it is not

    /// Not a sale. Nothing left the shelf and the shop earned nothing.
    @Test func aLoanIsNotASaleAndEarnsNothing() throws {
        let store = makeStore()
        let period = StatementPeriod.month(day)

        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        #expect(store.soldIn(period) == 0, "not takings")
        #expect(store.earningsIn(period).sold == 0, "not revenue")
        #expect(store.spentIn(period) == 0, "and not the owner's spending either")
        #expect(store.billCountIn(period) == 0, "no bill was written")
    }

    /// Not an expense. An expense is gone; this is expected back.
    @Test func lendingStandsApartFromSpending() throws {
        let store = makeStore()
        let period = StatementPeriod.month(day)
        _ = store.addExpense(amount: 60, note: "Petrol", spentAt: day)
        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        #expect(store.spentIn(period) == 60)
        #expect(store.lentIn(period) == 400)
    }

    /// Neither zero nor a negative figure is a loan. A negative one is a payment.
    @Test func aLoanHasToBeMoneyAndHasToBeForSomebody() throws {
        let store = makeStore()

        #expect(store.recordLoan(customerKey: "ahmed", amount: 0, lentAt: day) == nil)
        #expect(store.recordLoan(customerKey: "ahmed", amount: -400, lentAt: day) == nil)
        #expect(store.recordLoan(customerKey: "", amount: 400, lentAt: day) == nil)
        #expect(store.loans.isEmpty)
    }

    // MARK: On the statement

    /// A loan charges the account and settles nothing — the mirror of a payment,
    /// which settles and charges nothing.
    @Test func theStatementShowsALoanAsAChargeThatSettlesNothing() throws {
        let store = makeStore()
        _ = store.saveBill(customer: "Ahmed", paid: 0, amount: 800, createdAt: day)
        _ = store.recordLoan(
            customerKey: "ahmed", amount: 400, lentAt: day.addingTimeInterval(3_600)
        )

        let statement = try #require(
            store.statement(forCustomer: "ahmed", period: .month(day))
        )
        let loan = try #require(statement.entries.first {
            if case .loan = $0 { return true } else { return false }
        })

        #expect(loan.charge == 400)
        #expect(loan.settledAtOnce == 0)
        #expect(statement.closingBalance == 1_200)
    }

    /// **Billed says what was billed.** Eight hundred of goods and four hundred
    /// lent is not "billed 1,200" — the customer never got an invoice for the
    /// cash, and would be right to query one.
    ///
    /// This is why a loan is its own `Entry.Kind` rather than another trade entry.
    @Test func aLoanIsKeptOutOfWhatWasBilled() throws {
        let store = makeStore()
        _ = store.saveBill(customer: "Ahmed", paid: 0, amount: 800, createdAt: day)
        _ = store.recordLoan(
            customerKey: "ahmed", amount: 400, lentAt: day.addingTimeInterval(3_600)
        )

        let statement = try #require(
            store.statement(forCustomer: "ahmed", period: .month(day))
        )

        #expect(statement.billed == 800, "goods only")
        #expect(statement.lent == 400, "and the cash on its own line")
        #expect(statement.received == 0)
    }

    /// No number to print, because a loan comes out of no numbered book.
    @Test func aLoanRowIsNamedRatherThanNumbered() throws {
        let store = makeStore()
        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)
        let statement = try #require(
            store.statement(forCustomer: "ahmed", period: .month(day))
        )

        let row = try #require(statement.entries.first)
        #expect(StatementDocument.reference(row, english) == "Loan")
    }

    // MARK: On the day

    /// Money out of the cash box on the day it left it, however certain the owner
    /// is of getting it back. The day's net is what the box did.
    @Test func aLoanIsMoneyOutOnTheDayItWasHandedOver() throws {
        let store = makeStore()
        _ = store.recordPayment(customerKey: "ahmed", amount: 100, receivedAt: day)
        _ = store.recordLoan(
            customerKey: "ahmed", amount: 400, lentAt: day.addingTimeInterval(60)
        )

        let book = store.dayBook(day)

        #expect(book.entries.contains { $0.kind == .loan })
        #expect(book.moneyIn == 100)
        #expect(book.moneyOut == 400)
        #expect(book.net == -300)
    }

    // MARK: Correcting one

    @Test func aLoanCanBeCorrectedAndRemoved() throws {
        let store = makeStore()
        let loan = try #require(store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day))

        _ = store.updateLoan(id: loan.id, amount: 250, lentAt: day)
        #expect(store.customer(key: "ahmed")?.owed == 250)

        store.deleteLoan(id: loan.id)
        #expect(store.customer(key: "ahmed")?.owed == 0)
        #expect(store.loans.isEmpty)
    }

    // MARK: Across the wire

    /// A loan survives export and import, and the file says version 5.
    ///
    /// The bump is the point: a reader built before loans existed would drop them
    /// and show every borrower owing less than they do, so it must refuse the
    /// file instead. That is the credit-note rule pointing the other way.
    @Test func loansTravelInTheBackupWhichIsVersionFive() throws {
        let store = makeStore()
        _ = store.saveBill(customer: "Ahmed", paid: 0, amount: 800, createdAt: day)
        _ = store.recordLoan(
            customerKey: "ahmed", amount: 400, lentAt: day, note: "for the school fees"
        )

        let document = store.makeBackupDocument(at: day)
        #expect(document.version == 5)

        let restored = StockbookStore(repository: InMemoryRepository())
        restored.replaceEverything(with: try BackupService.decode(BackupService.encode(document)))

        let loan = try #require(restored.loans.first)
        #expect(restored.loans.count == 1)
        #expect(loan.amount == 400)
        #expect(loan.customerKey == "ahmed")
        #expect(loan.note == "for the school fees")
        #expect(
            restored.customer(key: "ahmed")?.owed == 1_200,
            "and the balance comes back with it"
        )
    }

    /// A file written before loans existed still loads.
    ///
    /// The Swift trap this exists for: a default does **not** make the
    /// synthesised decoder tolerate a missing key — it throws. `loans` is read
    /// with `decodeIfPresent` by hand in both `ShopState` and `BackupDocument`,
    /// and this is what proves it.
    ///
    /// Every other array is present, and that is not padding. `BackupDocument`'s
    /// decoder is strict on purpose about the keys that have always existed: a
    /// file with no `customers` or `purchases` is not one this app wrote, and
    /// refusing it is the right answer. Only the keys added after v1 are
    /// tolerant. The first draft of this test left them out and was rejected —
    /// correctly.
    @Test func aFileWrittenBeforeLoansExistedStillLoads() throws {
        let json = """
        {"version":4,"exportedAt":"2026-08-11T00:00:00Z","ownerName":"K",\
        "currencyCode":"SAR","products":[],"bills":[],"customers":[],\
        "payments":[],"suppliers":[],"purchases":[],"supplierPayments":[]}
        """
        let document = try BackupService.decode(Data(json.utf8))

        #expect(document.loans.isEmpty)
    }

    // MARK: Finding one

    @Test func aLoanCanBeSearchedForByTheBorrowersName() throws {
        let store = makeStore()
        _ = store.recordLoan(customerKey: "ahmed", amount: 400, lentAt: day)

        let hit = try #require(store.search("ahmed").first { $0.kind == .loan })

        #expect(hit.amount == 400)
        #expect(hit.who == "Ahmed")
        #expect(hit.reference == nil, "there is no number to show")
    }
}
